### PR TITLE
Update pystac.utils docs

### LINKED ISSUE #249: Document utils methods
Utils methods like `str_to_date` and `date_to_str` are useful to users when working with STAC. These are not currently documented in the API docs. This issue is to add them to the documentation page.

### PR DESCRIPTION
**Pull request by duckontheweb**
**Related Issue(s):**

- Closes #249

**Description:**

Updates docstrings with references to internal and external API docs and adds more detail to some of the docstrings.

**PR Checklist:**

- [x] Code is formatted (run `pre-commit run --all-files`)
- [x] Tests pass (run `scripts/test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] This PR maintains or improves overall codebase code coverage.
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/main/CHANGELOG.md). See [the docs](https://pystac.readthedocs.io/en/latest/contributing.html#changelog) for information about adding to the changelog.

### CODE REVIEW

File: pystac/utils.py [hunk 1]
```diff
@@ -38,19 +46,25 @@ def safe_urlparse(href: str) -> URLParseResult:
...
 
     @staticmethod
     def from_parsed_uri(parsed_uri: URLParseResult) -> "JoinType":
         """Determines the appropriate join type based on the scheme of the parsed
-        result."""
+        result.
+
+        Args:
+            parsed_uri (urllib.parse.ParseResult) : A named tuple representing the
+                parsed URI.
```
Comment (gadomski) [inline pystac/utils.py:66]: Should have a `Returns:` section to match the rest of the updates.

File: pystac/utils.py [hunk 2]
```diff
@@ -337,33 +380,86 @@ def extract_coords(coords: List[Union[List[float], List[List[Any]]]]) -> None:
 
 
 def map_opt(fn: Callable[[T], U], v: Optional[T]) -> Optional[U]:
-    """Maps the value of an option to another value, returning
-    None if the input option is None.
+    """Maps the value of an optional type to another value, returning
+    ``None`` if the input option is ``None``.
+
```
Comment (gadomski) [inline pystac/utils.py:385]: Should this have `Args:` and `Returns:`?

**Review by gadomski — CHANGES_REQUESTED** [review]
I got two batches of warnings when building locally, wanted to see if they were worth fixing with this PR. First:

```
/Users/gadomski/Code/pystac/pystac/extensions/datacube.py:docstring of pystac.extensions.datacube:3: WARNING: hardcoded link 'https://github.com/stac-extensions/datacube' could be replaced by an extlink (try using ':stac-ext:`datacube`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/eo.py:docstring of pystac.extensions.eo:3: WARNING: hardcoded link 'https://github.com/stac-extensions/eo' could be replaced by an extlink (try using ':stac-ext:`eo`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/file.py:docstring of pystac.extensions.file:3: WARNING: hardcoded link 'https://github.com/stac-extensions/file' could be replaced by an extlink (try using ':stac-ext:`file`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/item_assets.py:docstring of pystac.extensions.item_assets:3: WARNING: hardcoded link 'https://github.com/stac-extensions/item-assets' could be replaced by an extlink (try using ':stac-ext:`item-assets`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/label.py:docstring of pystac.extensions.label:3: WARNING: hardcoded link 'https://github.com/stac-extensions/label' could be replaced by an extlink (try using ':stac-ext:`label`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/pointcloud.py:docstring of pystac.extensions.pointcloud:3: WARNING: hardcoded link 'https://github.com/stac-extensions/pointcloud' could be replaced by an extlink (try using ':stac-ext:`pointcloud`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/projection.py:docstring of pystac.extensions.projection:3: WARNING: hardcoded link 'https://github.com/stac-extensions/projection' could be replaced by an extlink (try using ':stac-ext:`projection`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/raster.py:docstring of pystac.extensions.raster:3: WARNING: hardcoded link 'https://github.com/stac-extensions/raster' could be replaced by an extlink (try using ':stac-ext:`raster`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/raster.py:docstring of pystac.extensions.raster.RasterBand.apply:: WARNING: hardcoded link 'https://github.com/stac-extensions/raster/#data-types' could be replaced by an extlink (try using ':stac-ext:`raster/#data-types`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/raster.py:docstring of pystac.extensions.raster.RasterBand.create:: WARNING: hardcoded link 'https://github.com/stac-extensions/raster/#data-types' could be replaced by an extlink (try using ':stac-ext:`raster/#data-types`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/sar.py:docstring of pystac.extensions.sar:3: WARNING: hardcoded link 'https://github.com/stac-extensions/sar' could be replaced by an extlink (try using ':stac-ext:`sar`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/sat.py:docstring of pystac.extensions.sat:3: WARNING: hardcoded link 'https://github.com/stac-extensions/sat' could be replaced by an extlink (try using ':stac-ext:`sat`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/scientific.py:docstring of pystac.extensions.scientific:3: WARNING: hardcoded link 'https://github.com/stac-extensions/scientific' could be replaced by an extlink (try using ':stac-ext:`scientific`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/table.py:docstring of pystac.extensions.table:3: WARNING: hardcoded link 'https://github.com/stac-extensions/table' could be replaced by an extlink (try using ':stac-ext:`table`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/timestamps.py:docstring of pystac.extensions.timestamps:3: WARNING: hardcoded link 'https://github.com/stac-extensions/timestamps' could be replaced by an extlink (try using ':stac-ext:`timestamps`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/version.py:docstring of pystac.extensions.version:3: WARNING: hardcoded link 'https://github.com/stac-extensions/version' could be replaced by an extlink (try using ':stac-ext:`version`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/version.py:docstring of pystac.extensions.version.VersionRelType:3: WARNING: hardcoded link 'https://github.com/stac-extensions/version#relation-types' could be replaced by an extlink (try using ':stac-ext:`version#relation-types`' instead)
/Users/gadomski/Code/pystac/pystac/extensions/view.py:docstring of pystac.extensions.view:3: WARNING: hardcoded link 'https://github.com/stac-extensions/view' could be replaced by an extlink (try using ':stac-ext:`view`' instead)
/Users/gadomski/Code/pystac/docs/quickstart.ipynb:736: WARNING: hardcoded link 'https://github.com/stac-extensions/eo' could be replaced by an extlink (try using ':stac-ext:`eo`' instead)
/Users/gadomski/Code/pystac/docs/quickstart.ipynb:736: WARNING: hardcoded link 'https://github.com/stac-extensions/view' could be replaced by an extlink (try using ':stac-ext:`view`' instead)
/Users/gadomski/Code/pystac/docs/quickstart.ipynb:736: WARNING: hardcoded link 'https://github.com/stac-extensions/projection' could be replaced by an extlink (try using ':stac-ext:`projection`' instead)
/Users/gadomski/Code/pystac/docs/quickstart.ipynb:817: WARNING: hardcoded link 'https://github.com/stac-extensions/eo#item-properties-or-asset-fields' could be replaced by an extlink (try using ':stac-ext:`eo#item-properties-or-asset-fields`' instead)
```

Second:

```
/Users/gadomski/Code/pystac/pystac/serialization/identify.py:docstring of pystac.serialization.identify.STACJSONDescription.extensions:1: WARNING: duplicate object description of pystac.serialization.identify.STACJSONDescription.extensions, other instance in api/serialization/identify, use :noindex: for one of them
/Users/gadomski/Code/pystac/pystac/serialization/identify.py:docstring of pystac.serialization.identify.STACJSONDescription.object_type:1: WARNING: duplicate object description of pystac.serialization.identify.STACJSONDescription.object_type, other instance in api/serialization/identify, use :noindex: for one of them
/Users/gadomski/Code/pystac/pystac/serialization/identify.py:docstring of pystac.serialization.identify.STACJSONDescription.version_range:1: WARNING: duplicate object description of pystac.serialization.identify.STACJSONDescription.version_range, other instance in api/serialization/identify, use :noindex: for one of them
```

I also have a couple of inline comments re: quick fixups.

File: pystac/utils.py (hunk 1 above)
Comment (duckontheweb) [reply pystac/utils.py:66]: Good catch, added in b47dceb3da0952cd4148e710db98318ca9b17c73

File: pystac/utils.py (hunk 2 above)
Comment (duckontheweb) [reply pystac/utils.py:388]: Added in b47dceb3da0952cd4148e710db98318ca9b17c73

**Comment by duckontheweb** [comment]
> I got two batches of warnings when building locally, wanted to see if they were worth fixing with this PR. First:
> 
> ```
> /Users/gadomski/Code/pystac/pystac/extensions/datacube.py:docstring of pystac.extensions.datacube:3: WARNING: hardcoded link 'https://github.com/stac-extensions/datacube' could be replaced by an extlink (try using ':stac-ext:`datacube`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/eo.py:docstring of pystac.extensions.eo:3: WARNING: hardcoded link 'https://github.com/stac-extensions/eo' could be replaced by an extlink (try using ':stac-ext:`eo`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/file.py:docstring of pystac.extensions.file:3: WARNING: hardcoded link 'https://github.com/stac-extensions/file' could be replaced by an extlink (try using ':stac-ext:`file`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/item_assets.py:docstring of pystac.extensions.item_assets:3: WARNING: hardcoded link 'https://github.com/stac-extensions/item-assets' could be replaced by an extlink (try using ':stac-ext:`item-assets`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/label.py:docstring of pystac.extensions.label:3: WARNING: hardcoded link 'https://github.com/stac-extensions/label' could be replaced by an extlink (try using ':stac-ext:`label`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/pointcloud.py:docstring of pystac.extensions.pointcloud:3: WARNING: hardcoded link 'https://github.com/stac-extensions/pointcloud' could be replaced by an extlink (try using ':stac-ext:`pointcloud`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/projection.py:docstring of pystac.extensions.projection:3: WARNING: hardcoded link 'https://github.com/stac-extensions/projection' could be replaced by an extlink (try using ':stac-ext:`projection`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/raster.py:docstring of pystac.extensions.raster:3: WARNING: hardcoded link 'https://github.com/stac-extensions/raster' could be replaced by an extlink (try using ':stac-ext:`raster`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/raster.py:docstring of pystac.extensions.raster.RasterBand.apply:: WARNING: hardcoded link 'https://github.com/stac-extensions/raster/#data-types' could be replaced by an extlink (try using ':stac-ext:`raster/#data-types`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/raster.py:docstring of pystac.extensions.raster.RasterBand.create:: WARNING: hardcoded link 'https://github.com/stac-extensions/raster/#data-types' could be replaced by an extlink (try using ':stac-ext:`raster/#data-types`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/sar.py:docstring of pystac.extensions.sar:3: WARNING: hardcoded link 'https://github.com/stac-extensions/sar' could be replaced by an extlink (try using ':stac-ext:`sar`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/sat.py:docstring of pystac.extensions.sat:3: WARNING: hardcoded link 'https://github.com/stac-extensions/sat' could be replaced by an extlink (try using ':stac-ext:`sat`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/scientific.py:docstring of pystac.extensions.scientific:3: WARNING: hardcoded link 'https://github.com/stac-extensions/scientific' could be replaced by an extlink (try using ':stac-ext:`scientific`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/table.py:docstring of pystac.extensions.table:3: WARNING: hardcoded link 'https://github.com/stac-extensions/table' could be replaced by an extlink (try using ':stac-ext:`table`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/timestamps.py:docstring of pystac.extensions.timestamps:3: WARNING: hardcoded link 'https://github.com/stac-extensions/timestamps' could be replaced by an extlink (try using ':stac-ext:`timestamps`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/version.py:docstring of pystac.extensions.version:3: WARNING: hardcoded link 'https://github.com/stac-extensions/version' could be replaced by an extlink (try using ':stac-ext:`version`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/version.py:docstring of pystac.extensions.version.VersionRelType:3: WARNING: hardcoded link 'https://github.com/stac-extensions/version#relation-types' could be replaced by an extlink (try using ':stac-ext:`version#relation-types`' instead)
> /Users/gadomski/Code/pystac/pystac/extensions/view.py:docstring of pystac.extensions.view:3: WARNING: hardcoded link 'https://github.com/stac-extensions/view' could be replaced by an extlink (try using ':stac-ext:`view`' instead)
> /Users/gadomski/Code/pystac/docs/quickstart.ipynb:736: WARNING: hardcoded link 'https://github.com/stac-extensions/eo' could be replaced by an extlink (try using ':stac-ext:`eo`' instead)
> /Users/gadomski/Code/pystac/docs/quickstart.ipynb:736: WARNING: hardcoded link 'https://github.com/stac-extensions/view' could be replaced by an extlink (try using ':stac-ext:`view`' instead)
> /Users/gadomski/Code/pystac/docs/quickstart.ipynb:736: WARNING: hardcoded link 'https://github.com/stac-extensions/projection' could be replaced by an extlink (try using ':stac-ext:`projection`' instead)
> /Users/gadomski/Code/pystac/docs/quickstart.ipynb:817: WARNING: hardcoded link 'https://github.com/stac-extensions/eo#item-properties-or-asset-fields' could be replaced by an extlink (try using ':stac-ext:`eo#item-properties-or-asset-fields`' instead)
> ```

This first batch of warnings was due to our use of hard-coded links in a lot of the extension modules (e.g. [here](https://github.com/stac-utils/pystac/blob/ca49adb92d591fa81d8aa76f0580ee0faaa7364b/pystac/extensions/datacube.py#L3)) and in some of our tutorial notebooks. 6fc5008b9ca746087e1faf1c75c5b36ddda56d8e replaces the hard-coded links in the source code with `:stac-ext:` extlink references. The links in the tutorial notebooks are a little trickier because we provide direct links to these notebooks in our docs (see the "GitHub Version" links to the tutorials [here](https://pystac.readthedocs.io/en/latest/tutorials.html)), so the `:stac-ext:` reference won't render properly as a link in those notebooks. Using Markdown syntax for the links results in the same warnings (see [sphinx-doc/sphinx#10012(comment)](https://github.com/sphinx-doc/sphinx/issues/10112#issuecomment-1042391115)). 
I suppressed these warnings from the `extlink` extension in c39e43b714258fc2d3bdb887fd25203810da8f23, but I'm not sure if that's really the best approach since it suppresses *all* warnings coming out of that extension. We could also live with the remaining warnings for now until sphinx-doc/sphinx#10012 is resolved, I'm fine with either approach.

> Second:
> 
> ```
> /Users/gadomski/Code/pystac/pystac/serialization/identify.py:docstring of pystac.serialization.identify.STACJSONDescription.extensions:1: WARNING: duplicate object description of pystac.serialization.identify.STACJSONDescription.extensions, other instance in api/serialization/identify, use :noindex: for one of them
> /Users/gadomski/Code/pystac/pystac/serialization/identify.py:docstring of pystac.serialization.identify.STACJSONDescription.object_type:1: WARNING: duplicate object description of pystac.serialization.identify.STACJSONDescription.object_type, other instance in api/serialization/identify, use :noindex: for one of them
> /Users/gadomski/Code/pystac/pystac/serialization/identify.py:docstring of pystac.serialization.identify.STACJSONDescription.version_range:1: WARNING: duplicate object description of pystac.serialization.identify.STACJSONDescription.version_range, other instance in api/serialization/identify, use :noindex: for one of them
> ```
> 
> I also have a couple of inline comments re: quick fixups.

This batch of warnings is fixed by 3d75e7a6c3eda7baa3e1d13fb0153098dd40812c

**Comment by duckontheweb** [comment]
I had to configure `mypy` to ignore missing `sphinx.util` type stubs to satisfy the CI even though I was unable to reproduce this issue locally.

**Review by gadomski — APPROVED** [review]

**Pull request merged**